### ISSUE TITLE
nfl-oracle: production model's win-capability has never been backtested; naive baseline captures only 65% of hindsight-best vs field's 97.5%

### ISSUE THREAD

**Issue by cheeksmagunda** [cheeksmagunda/sports#280]
## Context

Investigating "why hasn't the model been capable of winning a draft" today, prompted by an operator request. Two things exist already in the codebase but neither has ever been connected to a real, recorded answer:

1. `nfl_oracle.replay.harness` analyzes the **field's** actual visible top-20 entries against hindsight-best (STATUS.md: winners captured a mean 79.9% of hindsight-best across all 91 finalized Corpus C contests; 97.5% in the four zero-boost week-1 contests specifically).
2. `nfl_oracle.replay.backtest.backtest_zero_boost_projection_from_history_index` replays a **naive EWMA-plus-position-prior projection** (no opponent, no injury/starter status, no odds, no weather, no card boost since it's zero-boost only) against Corpus G player-game history and reports a capture ratio. This function is exercised only by `tests/unit/test_replay_backtest.py` with synthetic 6-player fixtures. It has never been run against the real archive, and no result from it appears anywhere in STATUS.md.

## What I did

Ran #2 against the real local `data/raw/corpus_g` (37,710 rows, 2,392 distinct players, 2024-2025 seasons):

\`\`\`
games backtested: 569
mean capture ratio: 0.652 (65.2%)
median capture ratio: 0.663
min capture ratio: 0.151
max capture ratio: 0.979
\`\`\`

## The gap

The naive baseline captures **65.2%** mean of hindsight-best. Real human winners in the *same zero-boost regime* (the direct rule analog for this comparison per STATUS.md) captured **97.5%** mean. That's a roughly 32-point gap on a metric that's otherwise apples-to-apples (same scoring law, same slot structure, same zero-boost assumption).

## What this does NOT show

This is a real, previously-unrecorded, concerning number, but it measures the wrong model. `backtest_zero_boost_projection_from_history_index`'s EWMA-plus-position-prior projection is **not** the production prediction pipeline. Production (`nfl_oracle.recommendations.cli._model_bundle`) uses `fit_model` over `enrich_historical_rows`/`attach_enrichment`, which pulls in opponent-defense priors and other context features this backtest module never sees. **The actual production model has never been backtested against real historical outcomes at all** -- there is no harness that replays the full feature-enriched pipeline against Corpus G/C and reports a comparable capture ratio. That is the real gap, and it's a bigger unknown than the 65.2% number above.

Also ruled out, so as not to be re-investigated: the optimizer itself (`nfl_oracle.recommendations.optimizer._exact_search`) already solves the exact MILP joint player-and-slot assignment problem when scipy is available (confirmed available and active in this environment) -- it does not merely sort by projected value and is not leaving the "94% of the field fails to optimize slot order" edge on the table STATUS.md describes; that finding is about the field's play, not ours. Separately, production deliberately maximizes expected Total Value, not P(top 10%) or field win-rate (`simulation_metrics_are_diagnostics_and_do_not_reassign_the_frozen_lineup`) -- a real, documented design choice with probabilistic consequences in a top-heavy payout structure, not a bug, and not something to flip without the backtested evidence #37/#38 already ask for.

## Suggested next step

Build a backtest harness that replays the actual production feature pipeline (not the EWMA baseline) against Corpus G, chronologically, with the same capture-ratio metric, so this question has a real answer instead of an absent one. That is real, scoped engineering work, not a same-day fix.

## Not done today

No production model/optimizer/objective code was changed. The evidence above doesn't clear a "high confidence" bar for any specific code change to the real production model; changing it on the naive baseline's result would misattribute the naive baseline's known weaknesses (missing context features) to the production model, which hasn't been measured.

**Comment by cheeksmagunda** [comment]
## Update: ran the existing walk-forward eval CLI against real data, and traced production's own model-selection logic

Found `nfl_oracle.baselines.eval_cli` / `walk_forward.py` (compares classical baselines vs a ridge model, "feature_ridge") -- built, documented (\`make walk-forward-report\`), but like the backtest in the original report above, only ever run against a 21-label fixture. Ran it against the real full local Corpus G (47,206 labels, 2024-2025):

\`\`\`
rank[position_median]: n=23955 mae=0.7832 rmse=1.2466 bias=-0.3409
rank[feature_ridge]:   n=23955 mae=0.7987 rmse=1.1294 bias=0.0740
rank[player_mean]:     n=23955 mae=0.7987 rmse=1.1294 bias=0.0740
rank[position_mean]:   n=23955 mae=0.8578 rmse=1.1944 bias=0.0428
rank[global_mean]:     n=23955 mae=0.9347 rmse=1.2711 bias=0.0450
feature_ridge_vs_best_baseline: best=position_median delta_mae=+0.0155, ridge_better=False
\`\`\`

**feature_ridge does not beat plain per-position median on mean absolute error, at real scale, not just the fixture.** But it isn't simply worse either: position_median's bias is -0.341 (systematically underpredicts by nearly a third of a point on average) versus feature_ridge's 0.074, and feature_ridge's RMSE (1.129) beats position_median's (1.247). Given contest 839's actual winning margin was 0.29 points (STATUS.md, "Corpus C landed"), a -0.34 systematic bias is not a small thing to wave off even though its point-estimate MAE looks fine.

**Then traced whether this is what production actually uses.** It is not, directly -- \`nfl_oracle.recommendations.model.fit_model\` (the real production trainer, feeding \`RatingModel\`, used by \`_model_bundle\` in \`cli.py\`) is a separate ridge implementation from \`baselines.value_model.FeatureDrivenValueModel\`/\`feature_ridge\` above. But \`fit_model\` does its own internal chronological 80/20 holdout comparison against \`player_prior\`/\`position_mean\`/\`global_mean\`, and then this line overrides that comparison:

\`\`\`python
selected_estimator: EstimatorName = "ridge" if names else holdout_winner
\`\`\`

Whenever ANY context feature exists in training data, production forces \`ridge\`, full stop -- regardless of whether a baseline actually won the internal holdout MAE comparison it just computed. This is deliberate, not an accident: it is exactly what issue #212 (closed) asked for and verified as done ("Done when: New force retrain ... activates \`selected_estimator=ridge\` with nonzero context coefs"). #212's stated reason: \`position_mean\`/\`global_mean\`/\`player_prior\` selected-coefficient vectors literally zero out every context feature (pace, defense, matchup, depth), so letting a baseline win on one holdout split would silently and permanently drop that signal from future predictions, which #212 judged worse than accepting a slightly higher near-term holdout MAE.

**I am not reverting #212's decision.** The MAE numbers above are consistent with #212's own tradeoff (ridge costs a little MAE to keep context features alive) rather than contradicting it -- and #212's rationale about context-feature collapse still holds regardless of this new number. Unwinding a closed, verified, deliberately-reasoned decision on the strength of one new MAE comparison, without also weighing bias/RMSE (which favor ridge) or re-litigating #212's collapse concern, is exactly the kind of unvalidated change #37/#38 already warn against.

## What this actually establishes, precisely

- The production model's real predictive performance, and specifically whether the #212 tradeoff (small MAE cost for context-feature retention) is still the right one at current data volume, has a real, previously absent number now: **+0.0155 MAE cost, -0.267 bias improvement, -0.117 RMSE improvement**, versus the best simple baseline, at real 2024-2025 scale.
- Whether *production's* `fit_model` (not `feature_ridge`) shows the same pattern is still unmeasured -- they are different ridge implementations over different feature sets, and I have not built a harness for `fit_model`/`RatingModel` specifically. That remains the actual gap identified in the original report above.
- No code changed. No model/optimizer/selection logic touched. This is evidence, not a fix, and I'm not presenting it as one.

**Comment by cheeksmagunda** [comment]
**Production pipeline now backtested on real data (2026-09-24).** New harness `nfl_oracle.replay.production_backtest` (`make -C nfl-oracle production-backtest CONTEXT_SNAPSHOT=<path>`) replays `fit_model` over `attach_enrichment`/`enrich_historical_rows`, then `predict` and the production `optimize` (exact MILP), walk-forward over Corpus G (37,710 rows, 2024-2025). The model retrains weekly on labels final strictly before the week's first kickoff; the prior bank refreshes per slate. `LeakageError` fails the run if any row at or after the cutoff, or from the slate's own game, reaches training or priors (unit-tested). Code lands in a PR after tonight's games (not inside the T-40 window).

| | n | mean | median | min | max |
|---|---|---|---|---|---|
| production, per game | 554 | **67.9%** | 69.4% | 17.6% | 99.2% |
| naive EWMA, same 554 slates | 554 | 65.6% | | | |
| production, per ET day | 125 | 64.0% | 65.5% | 24.2% | 89.2% |
| production without context features | 554 | 51.7% | 52.2% | 6.5% | 91.7% |
| coefficients forced to player_mean_shrunk | 554 | 66.0% | 67.5% | 16.6% | 96.4% |
| coefficients forced to recent_mean_shrunk | 554 | 64.9% | 67.3% | 12.7% | 97.1% |

Production beats the naive baseline on 307 of 554 paired slates, only about 2.3 points on average, and is far below the field's 97.5% zero-boost / 79.9% all-contest numbers. Those use an easier denominator (visible ~45-player pool vs our full game pool), so the gap is not apples-to-apples, but it is clearly large. Context features are worth about 16 points (supports #212). No model variant tried beat production, so no production code changed.

Caveats: retrospective nflverse context (may include later revisions); no live injury/odds/weather inputs; DNP players removed from the pool (generous to both models); wall-clock training with a data cutoff because archive capture clocks are retrospective; Corpus C contest-pool comparison not yet built; 16 early slates skipped for insufficient history.

Leaving open: the harness answers "has the production model ever been backtested" (now yes), but the capability gap this issue is about remains. Next step: player selection, not the optimizer, is where the ~30 points go. Measure candidates against this harness (projection calibration by position, availability, and a Corpus C contest-pool replay for a like-for-like field comparison).
